Fixes #30750 - poweroff -r command for SSH finish

lzap opened the pull request:
Users must not do "reboot" command directly, we shall provide them a helper not to break SSH connection.

lzap committed:
Fixes #30750 - poweroff schedule command for SSH finish

ekohl reviewed (dismissed):
I'd also say that the PR title (`poweroff`) doesn't match the content (`reboot`). Perhaps we also want some test (at least kickstart?) that sets the parameter to true? Do we document these host parameters somewhere. Otherwise the code looks good and I have no objection to merging.

I notice that none of these use any whitespace control. Perhaps that's something we should look at in the future.

tbrisker commented:
snapshots need updating

lzap commented:
> I'd also say that the PR title (`poweroff`) doesn't match the content (`reboot`).

Fixed.

> Perhaps we also want some test (at least kickstart?) that sets the parameter to true?

Good idea, I have reviewed our templates and enabled some of them, snapshots now renders a little bit more, including puppet. This revealed some extra whitespace (particularly in the puppet snippet), I've fixed those. Also renabled ReX param, unfortunately I cannot enable Katello RHSM param, that would add significant amount of rendered ERB into snapshosts, the reason is some helpers which are only available in Katello.

> Do we document these host parameters somewhere.

We don't IIRC, however it's a different discussion. Many (if not most) params are "unless param is false" kind of statements (ability to turn off particular secion) and that works fine for snapshoting.

> I notice that none of these use any whitespace control. Perhaps that's something we should look at in the future.

Not sure what you mean, but I will fix whitespace introduced by this patch.

lzap commented:
Rebased.

ekohl commented on test/unit/foreman/renderer/snapshots/ProvisioningTemplate/cloud-init/CloudInit default.snap.txt:
```diff
@@ -70,10 +70,37 @@ runcmd:
+  
+  
+  [main]
+  vardir = /var/lib/puppet
+  logdir = /var/log/puppet
+  rundir = /var/run/puppet
+  ssldir = \$vardir/ssl
```
Historically we always put this here, but I'm wondering why we still do this. When you install Puppet it doesn't do this and doesn't need it either. In fact, it can only break things. Before adding Puppet configs in all snapshots and then removing them, should we instead first remove these?

lzap replied:
I am not sure what you mean, by my motivation is to enable as many arguments as I can so more content gets rendered. The main purpose of snapshosts is to render content so we know there are no syntax errors.

ekohl replied:
https://github.com/theforeman/foreman/pull/7979 is what I meant.

ekohl reviewed (commented):
> > Do we document these host parameters somewhere.
> 
> We don't IIRC, however it's a different discussion. Many (if not most) params are "unless param is false" kind of statements (ability to turn off particular secion) and that works fine for snapshoting.

At least for kickstart there's `This template accepts the following parameters:` as a documentation comment. It'd be good to include it there. This documentation comment isn't in every template, but please add it where it is present.

> > I notice that none of these use any whitespace control. Perhaps that's something we should look at in the future.
> 
> Not sure what you mean, but I will fix whitespace introduced by this patch.

It's a longer standing issue. If you look at the rendered template, there's always a lot of empty lines between parts. It doesn't change the effective result of rendered template, but I've noticed that in the past when I was looking at rendered previews.

lzap commented:
Ok so what is the status, do you want me to do some kind of documenting? I'd rather have a separate PR for that. 

I quickly grepped our codebase and we have the following list:

https://community.theforeman.org/t/appending-bootloader-arguments/20400/4

I'd rather make this a documentation effort rather than some comments in templates themselves.

ekohl commented:
> I'd rather make this a documentation effort rather than some comments in templates themselves.

That's how we do it now, let's at least keep them up to date since it's what users are used to.

lzap commented:
I am starting new effort in this regard: https://community.theforeman.org/t/help-needed-document-template-parameters-you-know/20595

I guess this PR can be merged now?

adamruzicka reviewed (approved):
LGTM

adamruzicka commented:
Thank you @lzap & @ekohl !

tbrisker commented:
I'm not sure if it's related but looks like snapshot tests are now failing, e.g.: https://ci.theforeman.org/job/test_develop_pr_core/6064/database=postgresql,ruby=2.5,slave=fast/#showFailuresLink

ezr-ondrej commented:
> I'm not sure if it's related but looks like snapshot tests are now failing, e.g.: https://ci.theforeman.org/job/test_develop_pr_core/6064/database=postgresql,ruby=2.5,slave=fast/#showFailuresLink

I believe it is:
```diff
--- expected
+++ actual
@@ -121,11 +121,14 @@
 
 cat > /etc/puppet/puppet.conf << EOF
+
+
 [main]
 vardir = /var/lib/puppet
 logdir = /var/log/puppet
 rundir = /var/run/puppet
 ssldir = \\$vardir/ssl
+digest_algorithm = sha256
 
 [agent]
 pluginsync      = true
```

lzap commented:
Weird, github automerged probably and there was a change somewhere else. Not sure, here is a PR: https://github.com/theforeman/foreman/pull/8016

lzap commented:
Ah I know what happened, since I called the command to generate them, I accidentally overwritten these, it passed here but not for core... https://github.com/theforeman/foreman/pull/7979
